src/packet.ml

```diff
@@ -247,21 +257,48 @@ let encode_protocol proto len =
       buf
   | `Udp -> Cstruct.empty
 
-let encode proto (key, p) =
-  let payload, len =
+let encode proto hmac_len
+    (key, (p : [< `Ack of header | `Control of operation * _ ])) =
+  let hdr = header p in
+  let len =
+    let id_arr_len = id_len * List.length hdr.ack_sequence_numbers in
+    protocol_len proto + 1 + hdr_len hmac_len + id_arr_len
+    + (if id_arr_len = 0 then 0 else 8)
+    +
     match p with
-    | `Ack ack -> encode_header ack
-    | `Control (_, control) -> encode_control control
-    | `Data d -> (d, Cstruct.length d)
+    | `Ack _ -> 0
+    | `Control (_, (_, _, payload)) -> 4 + Cstruct.length payload
   in
-  let op_buf =
-    let b = Cstruct.create 1 in
-    let op = op_key (operation p) key in
-    Cstruct.set_uint8 b 0 op;
-    b
+  let buf = Cstruct.create len in
+  set_protocol buf proto;
+  let op = op_key (operation p) key in
+  Cstruct.set_uint8 buf (protocol_len proto) op;
+  let op_buf = Cstruct.sub buf (protocol_len proto) 1 in
+  let to_encode = Cstruct.shift buf (protocol_len proto + 1) in
+  let () =
+    match p with
+    | `Ack ack -> ignore (encode_header hmac_len to_encode ack)
+    | `Control (_, control) -> encode_control hmac_len to_encode control
+  in
+  let feeder feed =
+    (* replay_id ++ timestamp *)
+    feed (Cstruct.sub to_encode (hmac_len + 8) (4 + 4));
+    feed op_buf;
+    (* local_session *)
+    feed (Cstruct.sub to_encode 0 8);
```

Comment:
the binding to op_buf can be removed.

We can `feed (Cstruct.sub buf (protocol_len proto) 9)` -- which includes the op_buf and the local session then :)